scripts/linking.py

```diff
@@ -1,86 +1,241 @@
+
+        # TODO: remove the following line and rebuild index
+        concept_aliases = [a if len(a) >= 2 else f'{a}00' for a in concept_aliases]
+
+        uml_concept_ids.extend([concept_id] * len(concept_aliases))
+        uml_concept_aliases.extend(concept_aliases)
+
+        if i % 1000000 == 0 and i > 0:
+            print(f'Processed {i} or {len(umls_concept_list)} concepts')
+
+    uml_concept_ids = np.array(uml_concept_ids)
+    uml_concept_aliases = np.array(uml_concept_aliases)
+    assert len(uml_concept_ids) == len(uml_concept_aliases)
+
+    if not os.path.isfile(tfidf_vectorizer_path):
+        print(f'No tfidf vectorizer on {tfidf_vectorizer_path}')
+        print(f'Fitting tfidf vectorizer on {len(uml_concept_aliases)} aliases')
+        tfidf_vectorizer = TfidfVectorizer(analyzer='char', ngram_range=(3, 4))
+        tfidf_vectorizer.fit(uml_concept_aliases)
+        print(f'Saving tfidf vectorizer to {tfidf_vectorizer_path}')
+        dump(tfidf_vectorizer, tfidf_vectorizer_path)
+    print(f'Loading tfidf vectorizer from {tfidf_vectorizer_path}')
+    tfidf_vectorizer = load(tfidf_vectorizer_path)
+    print(f'Vectorizing aliases ... ')
+    start_time = datetime.datetime.now()
+    uml_concept_aliase_tfidfs = tfidf_vectorizer.transform(uml_concept_aliases)
+    end_time = datetime.datetime.now()
+    total_time = (end_time - start_time)
+    print(f'Vectorizing aliases took {total_time.total_seconds()} seconds')
+
+    # find empty (all zeros) tfidf vectors
+    empty_tfidfs_boolean_flags = np.array(uml_concept_aliase_tfidfs.sum(axis=1) != 0).reshape(-1,)
+    deleted_aliases = uml_concept_aliases[empty_tfidfs_boolean_flags == False]
+    number_of_none_empty_tfidfs = len(deleted_aliases)
+    total_number_of_tfidfs = uml_concept_aliase_tfidfs.shape[0]
+    print(f'Deleting {number_of_none_empty_tfidfs}/{total_number_of_tfidfs} aliases because their tfidf is empty')
+
+    # remove empty tfidf vectors, otherwise nmslib will crashd
+    uml_concept_ids = uml_concept_ids[empty_tfidfs_boolean_flags]
+    uml_concept_aliases = uml_concept_aliases[empty_tfidfs_boolean_flags]
+    uml_concept_aliase_tfidfs = uml_concept_aliase_tfidfs[empty_tfidfs_boolean_flags]
+    print(deleted_aliases)
+    assert len(uml_concept_ids) == len(uml_concept_aliases)
+    assert len(uml_concept_ids) == uml_concept_aliase_tfidfs.shape[0]
+
+    # nmslib hyperparameters (very important)
+    M = 100
+    efC = 2000
+    num_threads = 60
+    efS = 1000
+    index_params = {'M': M, 'indexThreadQty': num_threads, 'efConstruction': efC, 'post' : 0}
+
+    if not os.path.isfile(ann_index_path):
+        print(f'No ann index on {ann_index_path}')
+        print(f'Fitting ann index on {len(uml_concept_aliases)} aliases (takes 2 hours)')
+
+        start_time = datetime.datetime.now()
+        ann_index = nmslib.init(method='hnsw', space='cosinesimil_sparse', data_type=nmslib.DataType.SPARSE_VECTOR)
```

Comment:
Could you add some comments discussing the performance tradeoffs you've researched here? It would be good to know for future users why these settings are 1) as they are and 2) why the hyperparams are important.